
message HardwareResourcesMessage
{
    required string slave_id =1;
    required string slave_uuid=8;
    required string slave_hostname=9;
    required CPUCollection cpu_collection = 2;
    required MemoryCollection mem_collection = 3;
    required GPUCollection    gpu_collection=4;
    required DiskCollection    disk_collection = 5;
    required PortCollection   port_collection = 6;
    required TLBCollection    tlb_collection = 7;
}

message HardwareResourcesMessages
{
    repeated HardwareResourcesMessage hardware_resources = 1;
}

message CPUCollection{
    repeated CPUInfo cpu_infos = 1;
    required int32 cpu_quantity= 2;
}

message CPUInfo{
    required int32 cpuID = 1;
    optional int32 coreID = 2;
    optional int32 physicalID = 3;
    optional int32 cpucores = 4;
    optional string modelname = 5;
    optional double cpuMHz = 6;
    optional double cpumaxMhz = 7;
    optional double cpuminMHz = 8;
    optional string l1dcache = 9;
    optional string l1icache = 10;
    optional string l2cache = 11;
    optional string l3cache = 12;
}

message TLBCollection{
    repeated TLBInfo tlb_infos = 1;
}

message TLBInfo{
    repeated string l1_data_tlb = 1;
    repeated string l1_instruction_tlb = 2;
    optional string l2_tlb = 3;
}

message MemoryCollection {
    required string max_size = 1;  // maximum capacity, not the actual memory size of current machine
    required int32 device_quantity = 2; // actual how many memory cards on the current machine
    repeated MemInfo mem_infos = 3;
}

message MemInfo {
    required string size = 1; // MB
    required string type = 2; // Type: DDR4
    required string speed = 3; // Speed: 2133 MHz
}

message GPUCollection {
    repeated GPUInfo gpu_infos = 1;
    required int32 gpu_quantity = 2;
}


message GPUInfo {
    required string description =2;
    required string product =3;
    required string vendor =4;
    optional int32 physical_id =5;
    optional string bus_info =6;
    optional string version =7;
    optional int32 width =8;           //unit 'bits'
    optional int32 clock =9;           //unit 'MHz'
    optional string capabilities =10;
    optional string configuration =11;
    optional string resources =12;
}


//{"name": "loop0", "maj:min": "7:0", "rm": "0", "size": "91M", "ro": "1", "type": "loop", "mountpoint": "/snap/core/6350"},
//{"name": "sda", "maj:min": "8:0", "rm": "0", "size": "1.8T", "ro": "0", "type": "disk", "mountpoint": null,
message DiskCollection{
    repeated DiskInfo disk_infos = 1;
    required int32 disk_quantity= 2;
}
message DiskInfo{
    enum DiskType {
        SSD = 0;
        HDD = 1;
    }
    required string   name = 1;  // "name": "sda"
    required string   size = 2; // "size": "1.8T"
    required DiskType type = 3; // "type": "disk"
    required string   disk_speed = 4;

}

message PortCollection {
    repeated PortInfo port_infos = 1;
    required int32 port_quantity = 2;
}
message PortInfo{
    optional string firstport =1;
    optional string lastport =2;

}
